pkg/controller/pxc/upgrade.go

```diff
@@ -119,7 +114,200 @@ func (r *ReconcilePerconaXtraDBCluster) updatePod(sfs api.StatefulApp, podSpec *
+		}
+	}
+
+	if err := r.waitPodRestart(updateRevision, pod, waitLimit); err != nil {
+		return fmt.Errorf("failed to wait pod: %v", err)
+	}
+
+	if err := r.waitPXCSynced(cr, pod.Status.PodIP, waitLimit); err != nil {
+		return fmt.Errorf("failed to wait pxc sync: %v", err)
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) getPrimaryPod(cr *api.PerconaXtraDBCluster) (string, error) {
+	user := "proxyadmin"
+	host := fmt.Sprintf("%s-proxysql-unready.%s", cr.ObjectMeta.Name, cr.Namespace)
+
+	var dbatabase queries.Database
+
+	for i := 0; ; i++ {
+		db, err := queries.New(r.client, cr.Namespace, cr.Spec.SecretsName, user, host, 6032)
+		if err != nil && i < int(cr.Spec.ProxySQL.Size) {
+			time.Sleep(time.Second)
+		} else if err != nil && i == int(cr.Spec.ProxySQL.Size) {
+			return "", err
+		} else {
+			dbatabase = db
+			break
+		}
+	}
+
+	defer dbatabase.Close()
+
+	return dbatabase.PrimaryHost()
+}
+
+func (r *ReconcilePerconaXtraDBCluster) waitPXCSynced(cr *api.PerconaXtraDBCluster, podIP string, waitLimit int) error {
+	user := "root"
+
+	dbatabase, err := queries.New(r.client, cr.Namespace, cr.Spec.SecretsName, user, podIP, 3306)
+	if err != nil {
+		return fmt.Errorf("failed to access PXC database: %v", err)
+	}
+
+	defer dbatabase.Close()
+
+	for i := 0; i < waitLimit; i++ {
+		state, err := dbatabase.WsrepLocalStateComment()
+		if err != nil {
+			return fmt.Errorf("failed to get wsrep local state: %v", err)
+		}
+
+		if state == "Synced" {
+			return nil
+		}
+
+		time.Sleep(time.Second * 1)
+	}
+
+	return fmt.Errorf("reach pod wait limit")
+}
+
+func (r *ReconcilePerconaXtraDBCluster) waitPodRestart(updateRevision string, pod *corev1.Pod, waitLimit int) error {
+	for i := 0; i < waitLimit; i++ {
+		time.Sleep(time.Second * 1)
+
+		err := r.client.Get(context.TODO(), types.NamespacedName{Name: pod.Name, Namespace: pod.Namespace}, pod)
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+
+		ready := false
+		for _, container := range pod.Status.ContainerStatuses {
+			if container.Name == "pxc" {
+				ready = container.Ready
+			}
+		}
+
+		if pod.Status.Phase == corev1.PodRunning && pod.ObjectMeta.Labels["controller-revision-hash"] == updateRevision && ready {
+			log.Info(fmt.Sprintf("pod %s started", pod.Name))
+			return nil
+		}
+	}
+
+	return fmt.Errorf("reach pod wait limit")
+}
+
+func isPXC(sfs api.StatefulApp) bool {
+	return sfs.Labels()["app.kubernetes.io/component"] == "pxc"
+}
+
+func (r *ReconcilePerconaXtraDBCluster) isBackupRunning(cr *api.PerconaXtraDBCluster) error {
+	bcpList := api.PerconaXtraDBClusterBackupList{}
+	if err := r.client.List(context.TODO(), &client.ListOptions{Namespace: cr.Namespace}, &bcpList); err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get backup object: %v", err)
+	}
+
+	for _, bcp := range bcpList.Items {
+		if bcp.Status.State == api.BackupRunning || bcp.Status.State == api.BackupStarting {
```

Comment:
we can extend backup status in future, so better will be 
```suggestion
		if bcp.Status.State != api.BackupSucceeded && bcp.Status.State != api.BackupFailed {
```